cfgov/archival/utils.py

```diff
@@ -0,0 +1,118 @@
+import json
+import logging
+
+from django.apps import apps
+from django.core.serializers.json import DjangoJSONEncoder
+from django.db.migrations.recorder import MigrationRecorder
+from django.utils import timezone
+
+from wagtail.coreutils import find_available_slug
+from wagtail.models import get_streamfield_names
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_last_migration(app):
+    Migration = MigrationRecorder.Migration
+    app_migrations = Migration.objects.filter(app=app).order_by("-applied")
+    last_migration = app_migrations.first()
+
+    if last_migration is None:
+        return ""
+
+    return last_migration.name
+
+
+def export_page(page):
+    """Get a copy of the page as JSON."""
+
+    # This includes its app_label, model, and the latest migration applied for
+    # the app, allowing any import to quickly verify whether import is
+    # possible (due to app or model availability and/or schema changes).
+    page_export = {
+        "app_label": page.content_type.app_label,
+        "model": page.content_type.model,
+        "last_migration": get_last_migration(page.content_type.app_label),
+        "exported_at": timezone.now().isoformat(),
+    }
+
+    # The serializable_data comes from django-modelcluster.
+    page_export["data"] = page.serializable_data()
+
+    # Convert streamfields, which serialize as strings containing JSON, back
+    # into objects. This will keep the final JSON cleaner, and not with
+    # JSON-in-a-string-in-JSON
+    streamfield_names = get_streamfield_names(page.specific.__class__)
+    for streamfield_name in streamfield_names:
+        page_export["data"][streamfield_name] = json.loads(
+            page_export["data"][streamfield_name]
+        )
+
+    # Strip the pk out of the data so that, on import, from_serializable_data()
+    # always creates a new object.
+    page_export["data"]["pk"] = None
+
+    # Dump the page data to JSON
+    page_json = json.dumps(
+        page_export, ensure_ascii=False, indent=4, cls=DjangoJSONEncoder
+    )
+
+    logger.info(f"Exported {page.slug} to JSON")
+    return page_json
+
+
+def import_page(parent_page, page_json, slug=None):
+    page_data = json.loads(page_json)
+
+    # Verify that we can load import the page
+    app_label = page_data["app_label"]
+    model_name = page_data["model"]
+    page_last_migration = page_data["last_migration"]
+
+    # Get the specific model that the imported page belongs to
+    try:
+        model = apps.get_model(app_label, model_name)
+    except LookupError as exc:
+        logger.error(
+            f"Unable to import page of type {app_label}.{model_name}; {exc}"
+        )
+        raise
+
+    # Compare the last migration
+    last_migration = get_last_migration(app_label)
+    if last_migration != page_last_migration:
```

Comment:
Another potential issue with using migrations as a compatibility check is that we may introduce migrations to a different app that nonetheless impact the page model being imported. For example, if we modify `v1.CFGOVPage`, that'll generate a migration in the `v1` app but not in the `ask_cfpb` app, even though the `ask_cfpb.AnswerLandingPage` model will be impacted.

Using migrations seems both not conservative enough (because it misses changes in other apps) but also potentially too conservative (a migration might be introduced in another model in the app, or to some non-breaking thing like help text). Is it too risky to remove this check entirely and just attempt the import, and (a) handle errors more nicely (as suggested above) and/or (b) let the user work out if things haven't been restored properly?

Reply:
Yeah.

The problem this is an attempt to solve is that you can write anything to a StreamField as JSON. but that doesn't mean it matches the current schema as Wagtail understands it. So you could import data with fields that no longer exist, and then when dumped into the edit, that data is just gone. 

Most of our content is in `content` fields, and if a descendent of `CFGOVPage` modifies its `content` field, it has to redefine it, and will generate a migration for it. You're correct that if `CFGOVPage.sidefoot` changes, trying to import a page of a type that doesn't  modify it won't have that change reflected.

I think there's a balance here between being feature-complete and usable enough to test the hypothesis that we'll need to restore pages this way. It's probably worth talking to @csebianlander and @sonnakim to discuss what the editor experience should be (we had talked about it, but I don't see it reflected in the partial GHE search results ATM).

We could also track migrations all the way down the inheritance chain, but I think at that point my preference would be to remove the migration check entirely and let users fend for themselves. Because

> Is it too risky to remove this check entirely

I think the risk is very low either way. Ultimately this is to enable the hypothesis-testing: do we need the ability to restore deleted content?